src/CSV.jl

```diff
@@ -364,141 +380,138 @@ function file(source,
+        rows, tapes, refs, typecodes = multithreadparse(typecodes, buf, datapos, len, options, rowsguess, pool, ncols, ignoreemptylines, typemap, limit, cmt, intsentinel, debug)
+        finalrows = sum(rows)
     else
-        tapelen = rowsguess * 2
-        tapes = Vector{UInt64}[Mmap.mmap(Vector{UInt64}, tapelen) for i = 1:ncols]
+        tapes, poslens = allocate(rowsguess, ncols, typecodes)
         refs = Vector{Dict{String, UInt64}}(undef, ncols)
         lastrefs = zeros(UInt64, ncols)
         t = Base.time()
-        rows, tapes = parsetape(Val(transpose), ignoreemptylines, ncols, gettypecodes(typemap), tapes, tapelen, buf, datapos, len, limit, cmt, positions, pool, refs, lastrefs, rowsguess, typecodes, debug, options, false)
+        rows, tapes = parsetape(Val(transpose), ignoreemptylines, ncols, gettypecodes(typemap), tapes, poslens, buf, datapos, len, limit, cmt, positions, pool, refs, lastrefs, rowsguess, typecodes, intsentinel, debug, options, false)
+        finalrows = rows
         debug && println("time for initial parsing to tape: $(Base.time() - t)")
     end
     for i = 1:ncols
         typecodes[i] &= ~USER
     end
     finaltypes = Type[TYPECODES[T] for T in typecodes]
     debug && println("types after parsing: $finaltypes, pool = $pool")
-    finalrefs = Vector{Vector{String}}(undef, ncols)
+    finalrefs = Vector{Union{Vector{String}, Nothing}}(undef, ncols)
     if pool > 0.0
         for i = 1:ncols
             if isassigned(refs, i)
                 finalrefs[i] = map(x->x[1], sort!(collect(refs[i]), by=x->x[2]))
-            elseif typecodes[i] == POOL || typecodes[i] == (POOL | MISSING)
+            elseif typebits(typecodes[i]) == POOL
                 # case where user manually specified types, but no rows were parsed
                 # so the refs never got initialized; initialize them here to empty
                 finalrefs[i] = Vector{String}[]
+            else
+                finalrefs[i] = nothing
             end
         end
+    else
+        fill!(finalrefs, nothing)
     end
-    return File(getname(source), names, finaltypes, rows, ncols, eq, categorical, finalrefs, buf, tapes)
+    if threaded === true
+        columns = AbstractVector[ApplyArray(vcat, (Column{_eltype(finaltypes[i]), finaltypes[i]}(tapes[j][i], rows[j], eq, categorical, finalrefs[i], buf, finaltypes[i] >: Int64 ? uint64(INT_SENTINELS[intsentinel.x]) : sentinelvalue(Base.nonmissingtype(finaltypes[i]))) for j = 1:Threads.nthreads())...) for i = 1:ncols]
+    else
+        columns = AbstractVector[Column{_eltype(finaltypes[i]), finaltypes[i]}(tapes[i], rows, eq, categorical, finalrefs[i], buf, finaltypes[i] >: Int64 ? uint64(INT_SENTINELS[intsentinel.x]) : sentinelvalue(Base.nonmissingtype(finaltypes[i]))) for i = 1:ncols]
+    end
+    lookup = Dict(k => v for (k, v) in zip(names, columns))
```

Comment:
I've seen a couple comments and we had at least one issue opened about the cost of `getproperty` on `CSV.Row` and `CSV.Row2`, so throwing a lookup Dict into `CSV.File` and `CSV.Rows` seems like not a high price to pay for faster property access later.